jolt-core/src/zkvm/ram/booleanity.rs

```diff
@@ -248,28 +256,42 @@ impl<F: JoltField> SumcheckInstance<F> for BooleanitySumcheck<F> {
                 prover_state.eq_r_r = prover_state.B.current_scalar;
 
                 // Compute H polynomials for each decomposed part
-                let trace = self.trace.as_ref().expect("Trace not set");
-                let memory_layout = self.memory_layout.as_ref().expect("Memory layout not set");
+                let addresses = self.addresses.as_ref().expect("Addresses not set");
 
                 let mut H_polys = Vec::with_capacity(self.d);
 
                 for i in 0..self.d {
-                    let H_vec: Vec<F> = trace
+                    let H_vec: Vec<F> = addresses
                         .par_iter()
-                        .map(|cycle| {
-                            remap_address(cycle.ram_access().address() as u64, memory_layout)
-                                .map_or(F::zero(), |address| {
-                                    // Get i-th address chunk
-                                    let address_i = (address >> (NUM_RA_I_VARS * (self.d - 1 - i)))
-                                        % (1 << NUM_RA_I_VARS);
-                                    prover_state.F[address_i as usize]
-                                })
+                        .map(|address_opt| {
+                            address_opt.map_or(F::zero(), |address| {
+                                // Get i-th address chunk
+                                let address_i = (address >> (NUM_RA_I_VARS * (self.d - 1 - i)))
+                                    % (1 << NUM_RA_I_VARS);
+                                prover_state.F.as_ref().unwrap()[address_i as usize]
+                            })
                         })
                         .collect();
                     H_polys.push(MultilinearPolynomial::from(H_vec));
                 }
 
                 prover_state.H = Some(H_polys);
+
+                // Drop G arrays and F array as they're no longer needed in phase 2
+                if let Some(g) = prover_state.G.take() {
+                    drop_in_background_thread(g);
+                }
+                if let Some(f) = prover_state.F.take() {
+                    drop_in_background_thread(f);
+                }
+
+                // Drop addresses and memory_layout as they're no longer needed in phase 2
+                if let Some(addresses) = self.addresses.take() {
+                    drop_in_background_thread(addresses);
+                }
+                if let Some(memory_layout) = self.memory_layout.take() {
```

Comment:
memory layout should be like a couple of integers, probably not worth bloating the codebase